File: HisabCore/Sources/HisabCore/HDFCXLSParser.swift
import Foundation

/// Parser for HDFC's legacy .xls statement export (CDF/BIFF8, read by MinimalXLS).
/// The sheet mirrors the .txt layout — one header row, one row per transaction with
/// separate Withdrawal/Deposit number cells — and shares HDFCStatementTable, so
/// identities match the PDF and TXT renditions exactly.
public struct HDFCXLSParser: StatementParser {
    public let source = Source.hdfc

    public init() {}

    public func canParse(data: Data, filename: String) -> Bool {
        guard let grid = try? MinimalXLS.cells(in: data) else { return false }
        let texts = grid.values.flatMap(\.values)
        return texts.contains("Chq./Ref.No.") && texts.contains { $0.contains("HDFC BANK") }
    }

    public func parse(data: Data, password: String?) throws -> ParsedDocument {
        let grid: [Int: [Int: String]]
        do {
            grid = try MinimalXLS.cells(in: data)
        } catch {
            throw ParseError.unrecognizedFormat
        }

        guard let headerRow = grid.first(where: { $0.value.values.contains("Chq./Ref.No.") }) else {
            throw ParseError.unrecognizedFormat
        }
        func column(_ title: String) -> Int? {
            headerRow.value.first { $0.value == title }?.key
        }
        guard let dateCol = column("Date"),
              let narrationCol = column("Narration"),
              let refCol = column("Chq./Ref.No."),
              let withdrawalCol = column("Withdrawal Amt."),
              let depositCol = column("Deposit Amt."),
              let balanceCol = column("Closing Balance") else {
            throw ParseError.unrecognizedFormat
        }

        var rows: [HDFCRow] = []
        for rowIndex in grid.keys.sorted() where rowIndex > headerRow.key {
            guard let row = grid[rowIndex], let date = row[dateCol],
                  date.wholeMatch(of: #/\d{2}/\d{2}/\d{2}/#) != nil else { continue }
            rows.append(HDFCRow(dateText: date,
                                narration: row[narrationCol] ?? "",
                                refText: row[refCol],
                                withdrawalText: row[withdrawalCol],
                                depositText: row[depositCol],
                                balanceText: row[balanceCol]))
        }

        let allText = grid.values.flatMap(\.values).joined(separator: "\n")
        return try HDFCStatementTable.parse(rows: rows,
                                            openingBalancePaise: nil,
                                            period: HDFCTXTParser.period(in: allText))
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/parsers/hdfc_xls_parser.dart
/// HDFC legacy .xls export (CDF/BIFF8 via MinimalXls). Port of
/// HDFCXLSParser.swift — shares HdfcStatementTable so identities match the
/// PDF and TXT renditions exactly.
library;

import '../domain.dart';
import '../generic_bank/minimal_xls.dart';
import 'hdfc_common.dart';

class HdfcXlsParser implements StatementParser {
  @override
  Source get source => Source.hdfc;

  const HdfcXlsParser();

  Map<int, Map<int, String>>? _grid(List<int> data) {
    try {
      return MinimalXls.cells(data);
    } on XlsCorruptException {
      return null;
    }
  }

  @override
  bool canParse(List<int> data, String filename) {
    final grid = _grid(data);
    if (grid == null) return false;
    final texts = grid.values.expand((r) => r.values);
    return texts.contains('Chq./Ref.No.') &&
        texts.any((t) => t.contains('HDFC BANK'));
  }

  @override
  ParsedDocument parse(List<int> data, {String? password}) {
    final grid = _grid(data);
    if (grid == null) throw const UnrecognizedFormatException();

    MapEntry<int, Map<int, String>>? headerRow;
    for (final entry in grid.entries) {
      if (entry.value.containsValue('Chq./Ref.No.')) {
        headerRow = entry;
        break;
      }
    }
    if (headerRow == null) throw const UnrecognizedFormatException();
    int? column(String title) {
      for (final entry in headerRow!.value.entries) {
        if (entry.value == title) return entry.key;
      }
      return null;
    }

    final dateCol = column('Date');
    final narrationCol = column('Narration');
    final refCol = column('Chq./Ref.No.');
    final withdrawalCol = column('Withdrawal Amt.');
    final depositCol = column('Deposit Amt.');
    final balanceCol = column('Closing Balance');
    if (dateCol == null ||
        narrationCol == null ||
        refCol == null ||
        withdrawalCol == null ||
        depositCol == null ||
        balanceCol == null) {
      throw const UnrecognizedFormatException();
    }

    final dateRegex = RegExp(r'^\d{2}/\d{2}/\d{2}$');
    final rows = <HdfcRow>[];
    final sortedKeys = grid.keys.toList()..sort();
    for (final rowIndex in sortedKeys) {
      if (rowIndex <= headerRow.key) continue;
      final row = grid[rowIndex];
      final date = row?[dateCol];
      if (row == null || date == null || !dateRegex.hasMatch(date)) continue;
      rows.add(HdfcRow(
        dateText: date,
        narration: row[narrationCol] ?? '',
        refText: row[refCol],
        withdrawalText: row[withdrawalCol],
        depositText: row[depositCol],
        balanceText: row[balanceCol],
      ));
    }

    final allText = grid.values.expand((r) => r.values).join('\n');
    return HdfcStatementTable.parse(
      rows: rows,
      openingBalancePaise: null,
      period: HdfcStatementTable.periodIn(allText),
    );
  }
}
